src/main/scala/firrtl/passes/MemUtils.scala

```diff
@@ -139,78 +134,88 @@ object MemPortUtils {
-    BundleType(
-      if (containsInfo(mem.info, "maskGran")) {
-        defaultSeq :+ {
-          if (getFillWMask(mem)) Field("mask", Default, flattenType(mem.dataType))
-          else Field("mask", Default, flattenType(create_mask(mem.dataType)))
-        }
-      }
-      else defaultSeq
-    )  
-  }
-  // TODO: Don't use create_mask???
+  def wPortToBundle(mem: DefMemory) = BundleType(
+    (defaultPortSeq(mem) :+ Field("data", Default, mem.dataType)) ++
+    (if (!containsInfo(mem.info, "maskGran")) Nil
+     else Seq(Field("mask", Default, createMask(mem.dataType))))
+  )
+  def wPortToFlattenBundle(mem: DefMemory) = BundleType(
+    (defaultPortSeq(mem) :+ Field("data", Default, flattenType(mem.dataType))) ++
+    (if (!containsInfo(mem.info, "maskGran")) Nil
+     else if (getFillWMask(mem)) Seq(Field("mask", Default, flattenType(mem.dataType)))
+     else Seq(Field("mask", Default, flattenType(createMask(mem.dataType)))))
+  )
+  // TODO: Don't use createMask???
 
-  def rwPortToBundle(mem: DefMemory) = {
-    val defaultSeq = defaultPortSeq(mem) ++ Seq(
-      Field("wmode", Default, UIntType(IntWidth(1))),
+  def rwPortToBundle(mem: DefMemory) = BundleType(
+    defaultPortSeq(mem) ++ Seq(
+      Field("wmode", Default, BoolType),
       Field("wdata", Default, mem.dataType),
       Field("rdata", Flip, mem.dataType)
+    ) ++ (if (!containsInfo(mem.info, "maskGran")) Nil
+     else Seq(Field("wmask", Default, createMask(mem.dataType)))
     )
-    BundleType(
-      if (containsInfo(mem.info, "maskGran")) defaultSeq :+ Field("wmask", Default, create_mask(mem.dataType))
-      else defaultSeq
-    )
-  }
+  )
 
-  def rwPortToFlattenBundle(mem: DefMemory) = {
-    val defaultSeq = defaultPortSeq(mem) ++ Seq(
+  def rwPortToFlattenBundle(mem: DefMemory) = BundleType(
+    defaultPortSeq(mem) ++ Seq(
       Field("wmode", Default, UIntType(IntWidth(1))),
       Field("wdata", Default, flattenType(mem.dataType)),
       Field("rdata", Flip, flattenType(mem.dataType))
-    )
-    BundleType(
-      if (containsInfo(mem.info, "maskGran")) {
-        defaultSeq :+ {
-          if (getFillWMask(mem)) Field("wmask", Default, flattenType(mem.dataType))
-          else Field("wmask", Default, flattenType(create_mask(mem.dataType)))
-        }
-      }
-      else defaultSeq
+    ) ++ (if (!containsInfo(mem.info, "maskGran")) Nil
+     else if (getFillWMask(mem)) Seq(Field("wmask", Default, flattenType(mem.dataType)))
+     else Seq(Field("wmask", Default, flattenType(createMask(mem.dataType))))
     )  
-  }
+  )
 
   def memToBundle(s: DefMemory) = BundleType(
-    s.readers.map(p => Field(p, Default, rPortToBundle(s))) ++
-      s.writers.map(p => Field(p, Default, wPortToBundle(s))) ++
-      s.readwriters.map(p => Field(p, Default, rwPortToBundle(s))))
+    s.readers.map(Field(_, Flip, rPortToBundle(s))) ++
+    s.writers.map(Field(_, Flip, wPortToBundle(s))) ++
+    s.readwriters.map(Field(_, Flip, rwPortToBundle(s))))
```

Comment:
This is also a bug fix / functional change and doesn't belong in a cleaning up commit